core/src/main/java/com/google/errorprone/bugpatterns/NamedLikeContextualKeyword.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2022 The Error Prone Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.errorprone.bugpatterns;
+
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+import static com.google.errorprone.matchers.Description.NO_MATCH;
+import static com.google.errorprone.matchers.Matchers.allOf;
+import static com.google.errorprone.matchers.Matchers.anyOf;
+import static com.google.errorprone.matchers.Matchers.methodIsConstructor;
+import static com.google.errorprone.matchers.Matchers.methodIsNamed;
+import static com.google.errorprone.matchers.Matchers.not;
+import static com.google.errorprone.util.ASTHelpers.streamSuperMethods;
+
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker.MethodTreeMatcher;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.MethodTree;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+
+/** Warns on classes or methods being named similarly to contextual keywords. */
+@BugPattern(
+    severity = WARNING,
+    summary = "Avoid naming of classes and methods that can be similar to contextual keywords")
+public final class NamedLikeContextualKeyword extends BugChecker implements MethodTreeMatcher {
+
+  // Refer to JLS 19 §3.9
+  private static final Matcher<MethodTree> DISALLOWED_CLASS_NAME_MATCHER =
+      allOf(
+          methodIsConstructor(),
+          anyOf(
+              methodIsNamed("exports"),
+              methodIsNamed("opens"),
+              methodIsNamed("requires"),
+              methodIsNamed("uses"),
+              methodIsNamed("module"),
+              methodIsNamed("permits"),
+              methodIsNamed("sealed"),
+              methodIsNamed("var"),
+              methodIsNamed("non-sealed"),
+              methodIsNamed("provides"),
+              methodIsNamed("to"),
+              methodIsNamed("with"),
+              methodIsNamed("open"),
+              methodIsNamed("record"),
+              methodIsNamed("transitive"),
+              methodIsNamed("yield")));
```

Comment:
We also have a copy of this list [here](https://github.com/PicnicSupermarket/error-prone-support/blob/f52a93cc4ede8f37577f0c02c7af4afcea954ef1/error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/JavaKeywords.java#L85-L102). I wonder whether there's value in exposing it as an `ImmutableSet<String>` in a utility class somewhere. :thinking: